File: docs/analytics.js
(() => {
  "use strict";

  const MEASUREMENT_ID = "G-D757SHS30N";
  const CONSENT_STORAGE_KEY = "j1_analytics_consent_v1";
  const LINKED_DOMAINS = [
    "akiokuyama.github.io",
    "j1league-score-prediction.streamlit.app",
  ];
  const VALID_CONSENT = new Set(["granted", "denied"]);
  const sentEventKeys = new Set();
  let configured = false;
  let tagRequested = false;

  window.dataLayer = window.dataLayer || [];
  window.gtag =
    window.gtag ||
    function gtag() {
      window.dataLayer.push(arguments);
    };

  window.gtag("consent", "default", {
    ad_storage: "denied",
    ad_user_data: "denied",
    ad_personalization: "denied",
    analytics_storage: "denied",
    wait_for_update: 500,
  });

  function readConsent() {
    try {
      const value = window.localStorage.getItem(CONSENT_STORAGE_KEY);
      return VALID_CONSENT.has(value) ? value : null;
    } catch {
      return null;
    }
  }

  function persistConsent(value) {
    try {
      window.localStorage.setItem(CONSENT_STORAGE_KEY, value);
    } catch {
      // 保存領域を利用できない場合も、現在のページでは選択を反映します。
    }
  }

  function applyConsent(value, { persist = true } = {}) {
    if (!VALID_CONSENT.has(value)) return;
    if (persist) persistConsent(value);
    window.gtag("consent", "update", {
      ad_storage: "denied",
      ad_user_data: "denied",
      ad_personalization: "denied",
      analytics_storage: value,
    });
    if (value === "granted") ensureGoogleTag();
    updateConsentUi(value);
    window.dispatchEvent(
      new CustomEvent("j1analyticsconsentchange", { detail: { consent: value } }),
    );
  }

  function ensureGoogleTag() {
    if (!tagRequested) {
      tagRequested = true;
      const script = document.createElement("script");
      script.async = true;
      script.src = `https://www.googletagmanager.com/gtag/js?id=${encodeURIComponent(MEASUREMENT_ID)}`;
      script.dataset.j1AnalyticsTag = MEASUREMENT_ID;
      document.head.appendChild(script);
    }
    if (!configured) {
      configured = true;
      window.gtag("js", new Date());
      window.gtag("set", "linker", { domains: LINKED_DOMAINS });
      window.gtag("config", MEASUREMENT_ID, {
        debug_mode: ["localhost", "127.0.0.1"].includes(window.location.hostname),
      });
    }
  }

  function sanitizeParameters(parameters) {
    return Object.fromEntries(
      Object.entries(parameters || {})
        .filter(([, value]) => ["string", "number", "boolean"].includes(typeof value))
        .map(([key, value]) => [key, typeof value === "string" ? value.slice(0, 100) : value]),
    );
  }

  function track(eventName, parameters = {}) {
    if (readConsent() !== "granted") return false;
    ensureGoogleTag();
    window.gtag("event", eventName, sanitizeParameters(parameters));
    return true;
  }

  function trackOnce(eventName, parameters = {}, eventKey = eventName) {
    if (sentEventKeys.has(eventKey)) return false;
    if (!track(eventName, parameters)) return false;
    sentEventKeys.add(eventKey);
    return true;
  }

  function consentLabel(value) {
    if (value === "granted") return "許可中";
    if (value === "denied") return "停止中";
    return "未選択";
  }

  function updateConsentUi(value = readConsent()) {
    document.querySelectorAll("[data-analytics-consent-status]").forEach((element) => {
      element.textContent = consentLabel(value);
    });
    document.querySelectorAll("[data-analytics-consent]").forEach((button) => {
      button.setAttribute("aria-pressed", String(button.dataset.analyticsConsent === value));
    });
    const banner = document.querySelector("[data-analytics-consent-banner]");
    if (banner) banner.hidden = value !== null;
  }

  function renderConsentBanner() {
    if (document.querySelector("[data-analytics-consent-banner]")) return;
    const banner = document.createElement("aside");
    banner.className = "analytics-consent-banner";
    banner.dataset.analyticsConsentBanner = "";
    banner.setAttribute("aria-label", "利用状況データの設定");
    banner.innerHTML = `
      <div class="analytics-consent-copy">
        <strong>利用状況データについて</strong>
        <p>
          サービス改善のため、同意いただいた場合のみGoogle Analyticsで画面表示や操作を計測します。
          個人を直接特定する情報は送信しません。
        </p>
      </div>
      <div class="analytics-consent-actions">
        <button type="button" class="analytics-consent-button analytics-consent-deny" data-analytics-consent="denied">
          今は許可しない
        </button>
        <button type="button" class="analytics-consent-button analytics-consent-allow" data-analytics-consent="granted">
          許可する
        </button>
      </div>
    `;
    document.body.appendChild(banner);
    updateConsentUi();
  }

  function bindConsentControls(root = document) {
    root.querySelectorAll("[data-analytics-consent]").forEach((button) => {
      if (button.dataset.analyticsConsentBound === "true") return;
      button.dataset.analyticsConsentBound = "true";
      button.addEventListener("click", () => {
        const previous = readConsent();
        const next = button.dataset.analyticsConsent;
        applyConsent(next);
        if (next === "granted") {
          track("analytics_consent_update", {
            consent_state: next,
            previous_state: previous || "unset",
          });
        }
      });
    });
  }

  function bindTrackedLinks(root = document) {
    root.querySelectorAll("[data-analytics-event]").forEach((link) => {
      if (link.dataset.analyticsEventBound === "true") return;
      link.dataset.analyticsEventBound = "true";
      link.addEventListener("click", () => {
        const parameters = {};
        for (const [key, value] of Object.entries(link.dataset)) {
          if (key.startsWith("analyticsParam")) {
            const rawName = key.slice("analyticsParam".length);
            const name = rawName
              .replace(/^[A-Z]/, (character) => character.toLowerCase())
              .replace(/[A-Z]/g, (character) => `_${character.toLowerCase()}`);
            parameters[name] = value;
          }
        }
        track(link.dataset.analyticsEvent, parameters);
      });
    });
  }

  function initialize() {
    const storedConsent = readConsent();
    if (storedConsent) applyConsent(storedConsent, { persist: false });
    renderConsentBanner();
    bindConsentControls();
    bindTrackedLinks();
    updateConsentUi(storedConsent);
  }

  window.J1Analytics = Object.freeze({
    measurementId: MEASUREMENT_ID,
    linkedDomains: [...LINKED_DOMAINS],
    getConsent: readConsent,
    setConsent: applyConsent,
    bindConsentControls,
    bindTrackedLinks,
    track,
    trackOnce,
    updateConsentUi,
  });

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", initialize, { once: true });
  } else {
    initialize();
  }
})();

File: app/utils/analytics.py
"""Consent-aware GA4 integration for the Streamlit application."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

import streamlit as st


MEASUREMENT_ID = "G-D757SHS30N"
CONSENT_STORAGE_KEY = "j1_analytics_consent_v1"
LINKED_DOMAINS = [
    "akiokuyama.github.io",
    "j1league-score-prediction.streamlit.app",
]

_ANALYTICS_BRIDGE_JS = r"""
export default function(component) {
    const { data, parentElement, setStateValue } = component;
    const measurementId = data.measurement_id;
    const storageKey = data.storage_key;
    const linkedDomains = Array.isArray(data.linked_domains) ? data.linked_domains : [];
    const events = Array.isArray(data.events) ? data.events : [];
    let storageError = null;

    window.__j1AnalyticsBridge = window.__j1AnalyticsBridge || {
        consentInitialized: false,
        tagRequested: false,
        configured: false,
        sentEventKeys: new Set(),
    };
    const bridge = window.__j1AnalyticsBridge;

    window.dataLayer = window.dataLayer || [];
    window.gtag = window.gtag || function() {
        window.dataLayer.push(arguments);
    };

    if (!bridge.consentInitialized) {
        bridge.consentInitialized = true;
        window.gtag("consent", "default", {
            ad_storage: "denied",
            ad_user_data: "denied",
            ad_personalization: "denied",
            analytics_storage: "denied",
            wait_for_update: 500,
        });
    }

    function readConsent() {
        try {
            const value = window.localStorage.getItem(storageKey);
            return value === "granted" || value === "denied" ? value : null;
        } catch (exception) {
            storageError = exception instanceof Error ? exception.message : String(exception);
            return null;
        }
    }

    function writeConsent(value) {
        try {
            window.localStorage.setItem(storageKey, value);
        } catch (exception) {
            storageError = exception instanceof Error ? exception.message : String(exception);
        }
    }

    function ensureTag() {
        if (!bridge.tagRequested) {
            bridge.tagRequested = true;
            const script = document.createElement("script");
            script.async = true;
            script.src = `https://www.googletagmanager.com/gtag/js?id=${encodeURIComponent(measurementId)}`;
            script.dataset.j1AnalyticsTag = measurementId;
            document.head.appendChild(script);
        }
        if (!bridge.configured) {
            bridge.configured = true;
            const debugMode = ["localhost", "127.0.0.1"].includes(window.location.hostname);
            window.gtag("js", new Date());
            window.gtag("set", "linker", { domains: linkedDomains });
            window.gtag("config", measurementId, { debug_mode: debugMode });
        }
    }

    let consent = readConsent();
    if (data.consent_action === "granted" || data.consent_action === "denied") {
        consent = data.consent_action;
        writeConsent(consent);
    }

    if (consent === "granted" || consent === "denied") {
        window.gtag("consent", "update", {
            ad_storage: "denied",
            ad_user_data: "denied",
            ad_personalization: "denied",
            analytics_storage: consent,
        });
    }

    if (consent === "granted") {
        ensureTag();
        for (const event of events) {
            const eventKey = String(event.key || "");
            if (!event.name || !eventKey || bridge.sentEventKeys.has(eventKey)) continue;
            window.gtag("event", event.name, event.parameters || {});
            bridge.sentEventKeys.add(eventKey);
        }
    }

    const snapshot = {
        loaded: true,
        consent: consent,
        storage_error: storageError,
        sent_event_count: bridge.sentEventKeys.size,
    };
    const serialized = JSON.stringify(snapshot);
    if (parentElement.dataset.analyticsSnapshot !== serialized) {
        parentElement.dataset.analyticsSnapshot = serialized;
        setStateValue("snapshot", snapshot);
    }
}
"""

_analytics_bridge = st.components.v2.component(
    "j1_ga4_analytics_bridge",
    js=_ANALYTICS_BRIDGE_JS,
    isolate_styles=False,
)


@dataclass(frozen=True)
class AnalyticsSnapshot:
    """State returned by the browser-side analytics bridge."""

    loaded: bool
    consent: Literal["granted", "denied"] | None = None
    storage_error: str | None = None
    sent_event_count: int = 0


def sync_analytics(
    *,
    component_key: str,
    consent_action: Literal["read", "granted", "denied"] = "read",
    events: list[dict[str, Any]] | None = None,
) -> AnalyticsSnapshot:
    """Synchronize consent and send a batch of deduplicated GA4 events."""

    result = _analytics_bridge(
        key=component_key,
        data={
            "measurement_id": MEASUREMENT_ID,
            "storage_key": CONSENT_STORAGE_KEY,
            "linked_domains": LINKED_DOMAINS,
            "consent_action": consent_action,
            "events": events or [],
        },
        default={
            "snapshot": {
                "loaded": False,
                "consent": None,
                "storage_error": None,
                "sent_event_count": 0,
            }
        },
        height=0,
        on_snapshot_change=lambda: None,
    )
    raw = getattr(result, "snapshot", None)
    if not isinstance(raw, dict):
        return AnalyticsSnapshot(loaded=False)
    consent = raw.get("consent")
    return AnalyticsSnapshot(
        loaded=bool(raw.get("loaded")),
        consent=consent if consent in {"granted", "denied"} else None,
        storage_error=str(raw["storage_error"]) if raw.get("storage_error") else None,
        sent_event_count=int(raw.get("sent_event_count") or 0),
    )


def queue_analytics_event(
    event_name: str,
    parameters: dict[str, str | int | float | bool] | None = None,
) -> None:
    """Queue one event for the next browser-bridge flush."""

    counter = int(st.session_state.get("_analytics_event_counter") or 0) + 1
    st.session_state._analytics_event_counter = counter
    sanitized = {
        str(key): value[:100] if isinstance(value, str) else value
        for key, value in (parameters or {}).items()
        if isinstance(value, (str, int, float, bool))
    }
    queue = st.session_state.setdefault("_analytics_event_queue", [])
    queue.append(
        {
            "name": event_name,
            "parameters": {"app_surface": "streamlit", **sanitized},
            "key": f"streamlit:{counter}:{event_name}",
        }
    )


def flush_analytics_events() -> AnalyticsSnapshot:
    """Send queued events and remove them from Streamlit session state."""

    events = st.session_state.pop("_analytics_event_queue", [])
    return sync_analytics(
        component_key="analytics_event_bridge",
        events=events if isinstance(events, list) else [],
    )


def track_state_change(
    state_key: str,
    value: Any,
    event_name: str,
    parameters: dict[str, str | int | float | bool] | None = None,
) -> None:
    """Queue an event only when an already-initialized widget value changes."""

    previous_key = f"_analytics_previous_{state_key}"
    if previous_key in st.session_state and st.session_state[previous_key] != value:
        queue_analytics_event(event_name, parameters)
    st.session_state[previous_key] = value
